// amazon-textract-response-parser/dist/es/api-models/document.js
export var ApiRelationshipType;
(function (ApiRelationshipType) {
    ApiRelationshipType["Child"] = "CHILD";
    ApiRelationshipType["ComplexFeatures"] = "COMPLEX_FEATURES";
    ApiRelationshipType["MergedCell"] = "MERGED_CELL";
    ApiRelationshipType["Value"] = "VALUE";
})(ApiRelationshipType || (ApiRelationshipType = {}));
export var ApiBlockType;
(function (ApiBlockType) {
    ApiBlockType["Cell"] = "CELL";
    ApiBlockType["KeyValueSet"] = "KEY_VALUE_SET";
    ApiBlockType["Line"] = "LINE";
    ApiBlockType["MergedCell"] = "MERGED_CELL";
    ApiBlockType["Page"] = "PAGE";
    ApiBlockType["SelectionElement"] = "SELECTION_ELEMENT";
    ApiBlockType["Table"] = "TABLE";
    ApiBlockType["Word"] = "WORD";
})(ApiBlockType || (ApiBlockType = {}));
export var ApiTextType;
(function (ApiTextType) {
    ApiTextType["Handwriting"] = "HANDWRITING";
    ApiTextType["Printed"] = "PRINTED";
})(ApiTextType || (ApiTextType = {}));
export var ApiKeyValueEntityType;
(function (ApiKeyValueEntityType) {
    ApiKeyValueEntityType["Key"] = "KEY";
    ApiKeyValueEntityType["Value"] = "VALUE";
})(ApiKeyValueEntityType || (ApiKeyValueEntityType = {}));
export var ApiSelectionStatus;
(function (ApiSelectionStatus) {
    ApiSelectionStatus["Selected"] = "SELECTED";
    ApiSelectionStatus["NotSelected"] = "NOT_SELECTED";
})(ApiSelectionStatus || (ApiSelectionStatus = {}));
