// amazon-textract-response-parser/dist/umd/api-models/document.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.ApiSelectionStatus = exports.ApiKeyValueEntityType = exports.ApiTextType = exports.ApiBlockType = exports.ApiRelationshipType = void 0;
    var ApiRelationshipType;
    (function (ApiRelationshipType) {
        ApiRelationshipType["Child"] = "CHILD";
        ApiRelationshipType["ComplexFeatures"] = "COMPLEX_FEATURES";
        ApiRelationshipType["MergedCell"] = "MERGED_CELL";
        ApiRelationshipType["Value"] = "VALUE";
    })(ApiRelationshipType = exports.ApiRelationshipType || (exports.ApiRelationshipType = {}));
    var ApiBlockType;
    (function (ApiBlockType) {
        ApiBlockType["Cell"] = "CELL";
        ApiBlockType["KeyValueSet"] = "KEY_VALUE_SET";
        ApiBlockType["Line"] = "LINE";
        ApiBlockType["MergedCell"] = "MERGED_CELL";
        ApiBlockType["Page"] = "PAGE";
        ApiBlockType["SelectionElement"] = "SELECTION_ELEMENT";
        ApiBlockType["Table"] = "TABLE";
        ApiBlockType["Word"] = "WORD";
    })(ApiBlockType = exports.ApiBlockType || (exports.ApiBlockType = {}));
    var ApiTextType;
    (function (ApiTextType) {
        ApiTextType["Handwriting"] = "HANDWRITING";
        ApiTextType["Printed"] = "PRINTED";
    })(ApiTextType = exports.ApiTextType || (exports.ApiTextType = {}));
    var ApiKeyValueEntityType;
    (function (ApiKeyValueEntityType) {
        ApiKeyValueEntityType["Key"] = "KEY";
        ApiKeyValueEntityType["Value"] = "VALUE";
    })(ApiKeyValueEntityType = exports.ApiKeyValueEntityType || (exports.ApiKeyValueEntityType = {}));
    var ApiSelectionStatus;
    (function (ApiSelectionStatus) {
        ApiSelectionStatus["Selected"] = "SELECTED";
        ApiSelectionStatus["NotSelected"] = "NOT_SELECTED";
    })(ApiSelectionStatus = exports.ApiSelectionStatus || (exports.ApiSelectionStatus = {}));
});
